# Data Cleaning

This notebook merges two raw, independently-sourced datasets into a single, clean, analysis-ready table:

- **World Bank (World Development Indicators)**: country-level development indicators (GDP, life expectancy, unemployment, inflation, population, FDI, CO2 emissions, poverty) across 25 years.
- **Heritage Foundation (Index of Economic Freedom)**: overall economic freedom score plus 12 sub-components (property rights, government integrity, trade freedom, etc.) by country and year.

The goal is not just to combine the two sources, but to "correct" real-world messiness, such as: inconsistent country names/codes, missing years, methodology changes over time, and structural mismatches between sources.

**Parts covered in this notebook:**
1. Imports, loading, and initial inspection
2. Country name/code standardisation
3. Reshaping from wide to long format
4. Merging both datasets
5. Deriving GDP per capita and continent
6. Handling missing values
7. Investigating outliers
8. Normalising numeric precision
9. Exporting the final clean dataset to `data/processed/`

---

<table align="left">
  <tr>
    <td>
      <a href="https://colab.research.google.com/github/ValentimPiazera/wb-economic-freedom/blob/main/notebooks/I-cleaning.ipynb" target="_parent">
        <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
      </a>
    </td>
    <td>
      <a target="_blank" href="https://kaggle.com/kernels/welcome?src=https://github.com/ValentimPiazera/wb-economic-freedom/blob/main/notebooks/I-cleaning.ipynb">
        <img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"/>
      </a>
    </td>
  </tr>
</table>

> **Note:** this notebook is written to run from the **repository root**, not from `notebooks/`.
> Locally, `.vscode/settings.json` already points VS Code's Jupyter root at the workspace folder;
> from a terminal, launch Jupyter from the repository root.
>
> If you're running on Google Colab or Kaggle, run the cell below first: it clones the repository
> so that the `src/` module and data files are available.
>
> On **Kaggle**, make sure Internet access is enabled: *Settings → Internet → On*.

In [1]:
# Setup: clone repo when running on Colab/Kaggle (skip if running locally!)
import os

IN_COLAB_OR_KAGGLE = (
    "google.colab" in str(get_ipython())
    or "kaggle" in os.environ.get("KAGGLE_URL_BASE", "").lower()
)

if IN_COLAB_OR_KAGGLE:
    if not os.path.exists("wb-economic-freedom"):
        !git clone https://github.com/ValentimPiazera/wb-economic-freedom.git
    # The repo root is the working directory, so `from src import cleaning`
    # and the `data/` paths below resolve the same way as they do locally.
    %cd wb-economic-freedom
    # Not preinstalled on Colab/Kaggle; used to derive the Continent column.
    !pip install -q pycountry_convert

## Part I | Imports, Loading, and Initial Inspection

Both raw sources are loaded and inspected as they arrive: shape, column names, data types, and the summary statistics that reveal which columns the World Bank export has delivered as strings.

Nothing is transformed in this section. It exists to establish what the two sources actually contain before any decision is made about them.

In [2]:
%load_ext autoreload
%autoreload 2

import pandas as pd

from src import cleaning

In [3]:
wb_df = pd.read_csv("data/raw/world_bank_indicators.csv")
ef_df = pd.read_csv("data/raw/economic_freedom_data.csv")

### World Bank — Initial Inspection

In [4]:
wb_df.head()

,Country Name,Country Code,Series Name,Series Code,2001 [YR2001],2002 [YR2002],2003 [YR2003],2004 [YR2004],2005 [YR2005],2006 [YR2006],...,2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022],2023 [YR2023],2024 [YR2024],2025 [YR2025]
0,Afghanistan,AFG,GDP (current US$),NY.GDP.MKTP.CD,2813571753.87253,3825701438.99963,4520946818.54581,5224896718.67782,6203256538.70967,6971758282.29351,...,18116572395.0772,18753456497.8159,18053222687.4126,18799444490.1128,19955929052.1496,14259995441.0759,14497243872.1337,17152234636.8715,17778508875.7396,..
1,Afghanistan,AFG,GDP growth (annual %),NY.GDP.MKTP.KD.ZG,-9.4319740700862,28.6000011706788,8.83227780288267,1.41411799339429,11.2297148272859,5.35740324748592,...,2.26031420279821,2.6470032027451,1.18922812944517,3.91160341625552,-2.35110067203466,-20.7388393676343,-6.24017199240269,2.26694373649188,1.87319280045979,..
2,Afghanistan,AFG,"Life expectancy at birth, total (years)",SP.DYN.LE00.IN,55.511,56.225,57.171,57.81,58.247,58.553,...,62.646,62.406,62.443,62.941,61.454,60.417,65.617,66.035,66.289,..
3,Afghanistan,AFG,"Unemployment, total (% of total labor force) (...",SL.UEM.TOTL.ZS,7.973,7.867,7.844,7.794,7.878,7.897,...,10.116,11.184,11.192,11.187,11.71,12.006,14.1,14.008,13.687,13.351
4,Afghanistan,AFG,"Inflation, consumer prices (annual %)",FP.CPI.TOTL.ZG,..,..,..,..,12.6862687216715,6.78459655001655,...,4.38389195513915,4.97595150553833,0.626149149168847,2.30237251516834,5.60188791482224,5.13320340824963,13.7121023720065,-4.64470870797775,-6.60118564073726,..


In [5]:
wb_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2175 entries, 0 to 2174
Data columns (total 29 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Country Name   2172 non-null   str  
 1   Country Code   2170 non-null   str  
 2   Series Name    2170 non-null   str  
 3   Series Code    2170 non-null   str  
 4   2001 [YR2001]  2170 non-null   str  
 5   2002 [YR2002]  2170 non-null   str  
 6   2003 [YR2003]  2170 non-null   str  
 7   2004 [YR2004]  2170 non-null   str  
 8   2005 [YR2005]  2170 non-null   str  
 9   2006 [YR2006]  2170 non-null   str  
 10  2007 [YR2007]  2170 non-null   str  
 11  2008 [YR2008]  2170 non-null   str  
 12  2009 [YR2009]  2170 non-null   str  
 13  2010 [YR2010]  2170 non-null   str  
 14  2011 [YR2011]  2170 non-null   str  
 15  2012 [YR2012]  2170 non-null   str  
 16  2013 [YR2013]  2170 non-null   str  
 17  2014 [YR2014]  2170 non-null   str  
 18  2015 [YR2015]  2170 non-null   str  
 19  2016 [YR2016]  21

In [6]:
wb_df.describe()

,Country Name,Country Code,Series Name,Series Code,2001 [YR2001],2002 [YR2002],2003 [YR2003],2004 [YR2004],2005 [YR2005],2006 [YR2006],...,2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022],2023 [YR2023],2024 [YR2024],2025 [YR2025]
count,2172,2170,2170,2170,2170,2170,2170,2170,2170,2170,...,2170,2170,2170,2170,2170,2170,2170,2170,2170,2170
unique,219,217,10,10,1759,1783,1780,1795,1804,1796,...,1816,1797,1808,1792,1791,1802,1788,1776,1703,947
top,Afghanistan,AFG,GDP (current US$),NY.GDP.MKTP.CD,..,..,..,..,..,..,...,..,..,..,..,..,..,..,..,..,..
freq,10,10,217,217,395,366,360,343,332,335,...,312,329,313,329,339,323,343,366,454,1223


### Economic Freedom Index — Initial Inspection

In [7]:
ef_df.head()

,Country,Index Year,Overall Score,Property Rights,Government Integrity,Judicial Effectiveness,Tax Burden,Government Spending,Fiscal Health,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom
0,Afghanistan,2026,NaN,3.6,13.5,0.0,92.0,90.8,98.4,33.7,44.4,84.9,NaN,NaN,NaN
1,Afghanistan,2025,NaN,7.4,14.1,2.7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,2024,NaN,4.9,18.1,4.9,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Afghanistan,2023,NaN,5.8,5.4,12.7,NaN,NaN,NaN,34.6,45.1,NaN,NaN,NaN,NaN
4,Afghanistan,2022,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
ef_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5888 entries, 0 to 5887
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Country                 5888 non-null   str    
 1   Index Year              5888 non-null   int64  
 2   Overall Score           5322 non-null   float64
 3   Property Rights         5388 non-null   float64
 4   Government Integrity    5404 non-null   float64
 5   Judicial Effectiveness  1832 non-null   float64
 6   Tax Burden              5347 non-null   float64
 7   Government Spending     5357 non-null   float64
 8   Fiscal Health           1800 non-null   float64
 9   Business Freedom        5386 non-null   float64
 10  Labor Freedom           3891 non-null   float64
 11  Monetary Freedom        5368 non-null   float64
 12  Trade Freedom           5351 non-null   float64
 13  Investment Freedom      5364 non-null   float64
 14  Financial Freedom       5342 non-null   float64
dty

In [9]:
ef_df.describe()

,Index Year,Overall Score,Property Rights,Government Integrity,Judicial Effectiveness,Tax Burden,Government Spending,Fiscal Health,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom
count,5888.000000,5322.000000,5388.000000,5404.000000,1832.000000,5347.000000,5357.000000,1800.000000,5386.000000,3891.000000,5368.000000,5351.000000,5364.000000,5342.000000
mean,2010.537874,59.598572,49.466425,41.702591,46.734061,74.124818,64.663114,63.732389,63.664835,59.441172,72.431967,69.501196,53.842282,49.543242
std,9.227285,11.665205,23.897350,22.586069,23.400168,15.111831,24.472821,31.838235,16.206571,15.197266,15.288104,15.299137,21.343599,19.699747
min,1995.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2003.000000,52.900000,30.000000,26.000000,29.275000,66.500000,51.000000,39.900000,55.000000,50.000000,69.400000,61.600000,40.000000,30.000000
50%,2011.000000,59.600000,50.000000,35.700000,43.600000,76.200000,70.800000,74.700000,64.800000,59.200000,75.600000,72.200000,55.000000,50.000000
75%,2019.000000,67.400000,70.000000,53.300000,62.600000,83.900000,84.100000,90.700000,73.600000,69.000000,80.800000,80.000000,70.000000,67.500000
max,2026.000000,90.500000,100.000000,100.000000,100.000000,100.000000,99.300000,100.000000,100.000000,100.000000,95.400000,95.000000,95.000000,90.000000


## Part II | Country Name and Code Standardisation

Before merging, both datasets need to represent countries the same way. World Bank and Heritage Foundation are independent sources, so naming conventions often diverge (for example: "Korea, Rep." vs "South Korea", or "Russian Federation" vs "Russia").

This section:
1. Compares the unique country names/codes in each dataset
2. Identifies mismatches (countries present in one source but not matched in the other)
3. Builds a standardised mapping to align both datasets on a common country identifier

In [10]:
wb_df["Country Name"].nunique()

219

In [11]:
ef_df["Country"].nunique()

186

In [12]:
only_in_wb, only_in_ef = cleaning.compare_countries(
    wb_df, ef_df, "Country Name", "Country", "World Bank", "Economic Freedom"
)

Only in World Bank: {'Venezuela, RB', 'Data from database: World Development Indicators', 'Gibraltar', 'Russian Federation', 'Puerto Rico (US)', 'Last Updated: 07/01/2026', 'Iran, Islamic Rep.', 'St. Kitts and Nevis', 'Nauru', 'Lao PDR', 'Sao Tome and Principe', 'Grenada', 'Syrian Arab Republic', 'Antigua and Barbuda', 'San Marino', 'British Virgin Islands', 'St. Martin (French part)', 'Channel Islands', 'Marshall Islands', 'Myanmar', 'Sint Maarten (Dutch part)', 'Palau', 'Viet Nam', 'St. Vincent and the Grenadines', 'Micronesia, Fed. Sts.', 'American Samoa', 'Bahamas, The', 'Hong Kong SAR, China', 'Northern Mariana Islands', 'Monaco', 'Cayman Islands', 'Egypt, Arab Rep.', 'Slovak Republic', 'Bermuda', 'Tuvalu', nan, 'Curacao', 'South Sudan', 'Greenland', 'Somalia, Fed. Rep.', 'Virgin Islands (U.S.)', 'New Caledonia', 'St. Lucia', 'Aruba', 'Korea, Rep.', 'Philippines', 'West Bank and Gaza', 'Faroe Islands', 'Macao SAR, China', 'Yemen, Rep.', 'Isle of Man', 'Gambia, The', "Korea, Dem. P

In [13]:
wb_df = cleaning.standardise_country_names(wb_df, "Country Name")

wb_df.head(10)

,Country Name,Country Code,Series Name,Series Code,2001 [YR2001],2002 [YR2002],2003 [YR2003],2004 [YR2004],2005 [YR2005],2006 [YR2006],...,2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022],2023 [YR2023],2024 [YR2024],2025 [YR2025]
0,Afghanistan,AFG,GDP (current US$),NY.GDP.MKTP.CD,2813571753.87253,3825701438.99963,4520946818.54581,5224896718.67782,6203256538.70967,6971758282.29351,...,18116572395.0772,18753456497.8159,18053222687.4126,18799444490.1128,19955929052.1496,14259995441.0759,14497243872.1337,17152234636.8715,17778508875.7396,..
1,Afghanistan,AFG,GDP growth (annual %),NY.GDP.MKTP.KD.ZG,-9.4319740700862,28.6000011706788,8.83227780288267,1.41411799339429,11.2297148272859,5.35740324748592,...,2.26031420279821,2.6470032027451,1.18922812944517,3.91160341625552,-2.35110067203466,-20.7388393676343,-6.24017199240269,2.26694373649188,1.87319280045979,..
2,Afghanistan,AFG,"Life expectancy at birth, total (years)",SP.DYN.LE00.IN,55.511,56.225,57.171,57.81,58.247,58.553,...,62.646,62.406,62.443,62.941,61.454,60.417,65.617,66.035,66.289,..
3,Afghanistan,AFG,"Unemployment, total (% of total labor force) (...",SL.UEM.TOTL.ZS,7.973,7.867,7.844,7.794,7.878,7.897,...,10.116,11.184,11.192,11.187,11.71,12.006,14.1,14.008,13.687,13.351
4,Afghanistan,AFG,"Inflation, consumer prices (annual %)",FP.CPI.TOTL.ZG,..,..,..,..,12.6862687216715,6.78459655001655,...,4.38389195513915,4.97595150553833,0.626149149168847,2.30237251516834,5.60188791482224,5.13320340824963,13.7121023720065,-4.64470870797775,-6.60118564073726,..
5,Afghanistan,AFG,"School enrollment, secondary (% gross)",SE.SEC.ENRR,14.0404100418091,..,13.959529876709,19.2143802642822,20.2735099792481,29.523889541626,...,53.4350395202637,55.5364303588867,57.3578491210938,..,..,..,..,59.6136016845703,..,..
6,Afghanistan,AFG,"Population, total",SP.POP.TOTL,20284307,21378117,22733049,23560654,24404567,25424094,...,34700612,35688935,36743039,37856121,39068979,40000412,40578842,41454761,42647492,43844111
7,Afghanistan,AFG,"Foreign direct investment, net inflows (% of GDP)",BX.KLT.DINV.WD.GD.ZS,0.0241685679088889,1.30694987042884,1.27849325196424,3.577104200584,4.36867310434287,3.41377297323201,...,0.516606084523093,0.274796791572807,0.661572217787196,0.124495985291258,0.0649939571808755,0.144466946606861,..,..,..,..
8,Afghanistan,AFG,Carbon dioxide (CO2) emissions (total) excludi...,EN.GHG.CO2.MT.CE.AR5,0.9402,0.9388,1.0172,0.9024,1.2763,1.4185,...,7.6117,10.3972,11.8099,12.1046,12.1436,12.4864,11.298,11.6486,12.0684,..
9,Afghanistan,AFG,Poverty headcount ratio at $3.00 a day (2021 P...,SI.POV.DDAY,..,..,..,..,..,..,...,..,..,..,..,..,..,..,..,..,..


In [14]:
only_in_wb, only_in_ef = cleaning.compare_countries(
    wb_df, ef_df, "Country Name", "Country", "World Bank", "Economic Freedom"
)

Only in World Bank: {'Data from database: World Development Indicators', 'Gibraltar', 'Last Updated: 07/01/2026', 'Puerto Rico (US)', 'St. Kitts and Nevis', 'Nauru', 'Grenada', 'Antigua and Barbuda', 'San Marino', 'British Virgin Islands', 'St. Martin (French part)', 'Channel Islands', 'Marshall Islands', 'Sint Maarten (Dutch part)', 'Palau', 'American Samoa', 'Northern Mariana Islands', 'Monaco', 'Cayman Islands', 'Bermuda', 'Tuvalu', nan, 'Curacao', 'South Sudan', 'Greenland', 'Virgin Islands (U.S.)', 'New Caledonia', 'West Bank and Gaza', 'Aruba', 'Faroe Islands', 'Isle of Man', 'Andorra', 'Guam', 'French Polynesia', 'Turks and Caicos Islands'}
Only in Economic Freedom: {'Taiwan'}


In [15]:
wb_df = cleaning.drop_unmatched_countries(
    wb_df, "Country Name", cleaning.ROWS_TO_DROP
)
wb_df = wb_df.dropna(subset=["Country Name"])

wb_df.head(10)

,Country Name,Country Code,Series Name,Series Code,2001 [YR2001],2002 [YR2002],2003 [YR2003],2004 [YR2004],2005 [YR2005],2006 [YR2006],...,2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022],2023 [YR2023],2024 [YR2024],2025 [YR2025]
0,Afghanistan,AFG,GDP (current US$),NY.GDP.MKTP.CD,2813571753.87253,3825701438.99963,4520946818.54581,5224896718.67782,6203256538.70967,6971758282.29351,...,18116572395.0772,18753456497.8159,18053222687.4126,18799444490.1128,19955929052.1496,14259995441.0759,14497243872.1337,17152234636.8715,17778508875.7396,..
1,Afghanistan,AFG,GDP growth (annual %),NY.GDP.MKTP.KD.ZG,-9.4319740700862,28.6000011706788,8.83227780288267,1.41411799339429,11.2297148272859,5.35740324748592,...,2.26031420279821,2.6470032027451,1.18922812944517,3.91160341625552,-2.35110067203466,-20.7388393676343,-6.24017199240269,2.26694373649188,1.87319280045979,..
2,Afghanistan,AFG,"Life expectancy at birth, total (years)",SP.DYN.LE00.IN,55.511,56.225,57.171,57.81,58.247,58.553,...,62.646,62.406,62.443,62.941,61.454,60.417,65.617,66.035,66.289,..
3,Afghanistan,AFG,"Unemployment, total (% of total labor force) (...",SL.UEM.TOTL.ZS,7.973,7.867,7.844,7.794,7.878,7.897,...,10.116,11.184,11.192,11.187,11.71,12.006,14.1,14.008,13.687,13.351
4,Afghanistan,AFG,"Inflation, consumer prices (annual %)",FP.CPI.TOTL.ZG,..,..,..,..,12.6862687216715,6.78459655001655,...,4.38389195513915,4.97595150553833,0.626149149168847,2.30237251516834,5.60188791482224,5.13320340824963,13.7121023720065,-4.64470870797775,-6.60118564073726,..
5,Afghanistan,AFG,"School enrollment, secondary (% gross)",SE.SEC.ENRR,14.0404100418091,..,13.959529876709,19.2143802642822,20.2735099792481,29.523889541626,...,53.4350395202637,55.5364303588867,57.3578491210938,..,..,..,..,59.6136016845703,..,..
6,Afghanistan,AFG,"Population, total",SP.POP.TOTL,20284307,21378117,22733049,23560654,24404567,25424094,...,34700612,35688935,36743039,37856121,39068979,40000412,40578842,41454761,42647492,43844111
7,Afghanistan,AFG,"Foreign direct investment, net inflows (% of GDP)",BX.KLT.DINV.WD.GD.ZS,0.0241685679088889,1.30694987042884,1.27849325196424,3.577104200584,4.36867310434287,3.41377297323201,...,0.516606084523093,0.274796791572807,0.661572217787196,0.124495985291258,0.0649939571808755,0.144466946606861,..,..,..,..
8,Afghanistan,AFG,Carbon dioxide (CO2) emissions (total) excludi...,EN.GHG.CO2.MT.CE.AR5,0.9402,0.9388,1.0172,0.9024,1.2763,1.4185,...,7.6117,10.3972,11.8099,12.1046,12.1436,12.4864,11.298,11.6486,12.0684,..
9,Afghanistan,AFG,Poverty headcount ratio at $3.00 a day (2021 P...,SI.POV.DDAY,..,..,..,..,..,..,...,..,..,..,..,..,..,..,..,..,..


In [16]:
ef_df = cleaning.drop_unmatched_countries(ef_df, "Country", ["Taiwan"])

In [17]:
only_in_wb, only_in_ef = cleaning.compare_countries(
    wb_df, ef_df, "Country Name", "Country", "World Bank", "Economic Freedom"
)

Only in World Bank: set()
Only in Economic Freedom: set()


### Dropping Unmatched Countries

After standardising country names, a small number of entries remained unmatched between the two datasets:

- **Junk rows:** The World Bank dataset contains metadata footer rows such as "Data from database..." and "Last Updated...", plus a stray `NaN` row. Since these are not countries, they were dropped.
- **Territories and dependencies:** Some countries are present in the World Bank dataset but not on the Economic Freedom Index (e.g. Greenland, Bermuda, Monaco, Puerto Rico). The Index only evaluates sovereign countries with sufficient economic data, so these have no possible match.
- **South Sudan:** A recognised sovereign country that is excluded from the Economic Freedom Index due to insufficient data availability.
- **Taiwan:** Present only in the Economic Freedom Index; it has no equivalent entry in the World Bank dataset due to its disputed sovereign status.

These rows were removed from both datasets, since they cannot be meaningfully merged. The final dataset therefore covers only countries present in **both** sources.

---

## Part III | Reshaping from Wide to Long Format

The World Bank dataset is currently in wide format: each row represents a country-indicator pair, with one column per year (e.g. `2001 [YR2001]`, `2002 [YR2002]`, ...). This structure is convenient for browsing but inconvenient for merging and analysis.

This section reshapes `wb_df` into long format: one row per country, per year, with each indicator as its own column. This is the structure required to merge with the Economic Freedom Index dataset and to support time-series analysis in the next notebook.

**Steps covered in this section:**
1. Melt the year columns into a single `Year` column
2. Pivot indicators from rows into columns
3. Clean up resulting column names and data types

In [18]:
id_vars = ["Country Name", "Country Code", "Series Name", "Series Code"]

wb_long = cleaning.melt_years(wb_df, id_vars)

wb_long.head()

,Country Name,Country Code,Series Name,Series Code,Year,Value
0,Afghanistan,AFG,GDP (current US$),NY.GDP.MKTP.CD,2001,2813571753.87253
1,Afghanistan,AFG,GDP growth (annual %),NY.GDP.MKTP.KD.ZG,2001,-9.4319740700862
2,Afghanistan,AFG,"Life expectancy at birth, total (years)",SP.DYN.LE00.IN,2001,55.511
3,Afghanistan,AFG,"Unemployment, total (% of total labor force) (...",SL.UEM.TOTL.ZS,2001,7.973
4,Afghanistan,AFG,"Inflation, consumer prices (annual %)",FP.CPI.TOTL.ZG,2001,..


In [19]:
wb_long.info()

<class 'pandas.DataFrame'>
RangeIndex: 46250 entries, 0 to 46249
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Country Name  46250 non-null  str  
 1   Country Code  46250 non-null  str  
 2   Series Name   46250 non-null  str  
 3   Series Code   46250 non-null  str  
 4   Year          46250 non-null  int64
 5   Value         46250 non-null  str  
dtypes: int64(1), str(5)
memory usage: 2.1 MB


In [20]:
wb_long.duplicated(subset=["Country Name", "Year", "Series Name"]).sum()

np.int64(0)

In [21]:
index_cols = ["Country Name", "Country Code", "Year"]

wb_final = cleaning.pivot_indicators(wb_long, index_cols)

wb_final.head()

,Country Name,Country Code,Year,Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),"Foreign direct investment, net inflows (% of GDP)",GDP (current US$),GDP growth (annual %),"Inflation, consumer prices (annual %)","Life expectancy at birth, total (years)","Population, total",Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population),"School enrollment, secondary (% gross)","Unemployment, total (% of total labor force) (modeled ILO estimate)"
0,Afghanistan,AFG,2001,0.9402,0.0241685679088889,2813571753.87253,-9.4319740700862,..,55.511,20284307,..,14.0404100418091,7.973
1,Afghanistan,AFG,2002,0.9388,1.30694987042884,3825701438.99963,28.6000011706788,..,56.225,21378117,..,..,7.867
2,Afghanistan,AFG,2003,1.0172,1.27849325196424,4520946818.54581,8.83227780288267,..,57.171,22733049,..,13.959529876709,7.844
3,Afghanistan,AFG,2004,0.9024,3.577104200584,5224896718.67782,1.41411799339429,..,57.81,23560654,..,19.2143802642822,7.794
4,Afghanistan,AFG,2005,1.2763,4.36867310434287,6203256538.70967,11.2297148272859,12.6862687216715,58.247,24404567,..,20.2735099792481,7.878


In [22]:
wb_final.shape

(4625, 13)

After melting and pivoting, `wb_final` now has one row per country-year combination, with each indicator as its own column: 4,625 rows (185 countries × 25 years) and 13 columns (3 identifier columns + 10 indicators). This is the structure required to merge with the Economic Freedom Index dataset.

---

## Part IV | Merging Both Datasets

With both datasets now in a compatible long format (one row per country-year), this section merges `wb_final` (World Bank indicators) with `ef_df` (Economic Freedom Index) into a single table.

An inner join is used on `Country` and `Year`, since unmatched countries were already excluded from both datasets in Part II. This keeps the merge explicit and avoids silently introducing `NaN` rows for any entries that were missed during standardisation.

**A caveat on the year alignment.** The two sources do not mean quite the same thing by "year". A World Bank indicator for year *N* describes calendar year *N*, whereas the Heritage Index published for year *N* is graded on data collected through roughly the first half of year *N-1* (which is why the raw Index file already contains an `Index Year` of 2026). Joining `Index Year` straight onto `Year` therefore pairs each freedom score with development outcomes measured slightly *after* the period the score describes.

This is kept as-is here, because it is the alignment a reader intuitively expects ("the 2020 Index next to 2020 outcomes") and because a one-year offset is small relative to how slowly these institutional scores move. It is recorded as an explicit decision rather than an oversight: the built-in lag is a choice, not a neutral default, and it is one of the reasons `II-analysis.ipynb` stops short of causal claims.

**Steps covered in this section:**
1. Merge both datasets on country and year
2. Drop the redundant country column
3. Verify the resulting shape and inspect for unexpected missing values

In [23]:
wb_final[["Country Name", "Year"]].dtypes


Country Name      str
Year            int64
dtype: object

In [24]:
ef_df[["Country", "Index Year"]].dtypes

Country         str
Index Year    int64
dtype: object

In [25]:
merged_df = wb_final.merge(
    ef_df,
    left_on=["Country Name", "Year"],
    right_on=["Country", "Index Year"],
    how="inner",
)

merged_df.head()

,Country Name,Country Code,Year,Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),"Foreign direct investment, net inflows (% of GDP)",GDP (current US$),GDP growth (annual %),"Inflation, consumer prices (annual %)","Life expectancy at birth, total (years)","Population, total",...,Judicial Effectiveness,Tax Burden,Government Spending,Fiscal Health,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom
0,Afghanistan,AFG,2001,0.9402,0.0241685679088889,2813571753.87253,-9.4319740700862,..,55.511,20284307,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Afghanistan,AFG,2002,0.9388,1.30694987042884,3825701438.99963,28.6000011706788,..,56.225,21378117,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,AFG,2003,1.0172,1.27849325196424,4520946818.54581,8.83227780288267,..,57.171,22733049,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Afghanistan,AFG,2004,0.9024,3.577104200584,5224896718.67782,1.41411799339429,..,57.81,23560654,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Afghanistan,AFG,2005,1.2763,4.36867310434287,6203256538.70967,11.2297148272859,12.6862687216715,58.247,24404567,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [26]:
merged_df.shape

(4581, 28)

In [27]:
# The merge is lossy: reconcile the result against the full country-year grid.
expected = wb_final["Country Name"].nunique() * wb_final["Year"].nunique()
missing_pairs = (
    wb_final[["Country Name", "Year"]]
    .merge(
        ef_df[["Country", "Index Year"]],
        left_on=["Country Name", "Year"],
        right_on=["Country", "Index Year"],
        how="left",
        indicator=True,
    )
    .query("_merge == 'left_only'")
)

print(f"Expected country-years: {expected}")
print(f"Merged rows: {len(merged_df)} ({expected - len(merged_df)} lost)")
missing_pairs["Country Name"].value_counts()

Expected country-years: 4625
Merged rows: 4581 (44 lost)


Country Name
Brunei Darussalam    13
Kosovo               11
Somalia              10
Hong Kong             5
Macau                 5
Name: count, dtype: int64

The merge keeps 4,581 of the 4,625 possible country-years, and the 44 missing pairs come from just five countries. This is not a standardisation failure: the Heritage Index simply does not grade every country every year, and each of these has a documented reason for its gap.

- **Brunei (13 years)** and **Kosovo (11)** were added to the Index only recently, so their early years do not exist.
- **Somalia (10)** was ungraded during the years in which state collapse left no reliable economic data to score.
- **Hong Kong (5)** and **Macau (5)** were graded for most of the period but removed from 2021 onwards, when the Index folded them into China on the grounds that they no longer set their own economic policy.

An inner join is the right treatment here: a row with no freedom score at all carries nothing for this project.

---

## Part V | Deriving GDP per Capita and Continent

The dataset currently includes total GDP (`GDP (current US$)`) and total population (`Population, total`), but not GDP per capita, a more meaningful measure of economic development when comparing countries of very different population sizes. It also has no geographic grouping yet, which will be useful for comparing countries by region in `II-analysis.ipynb`.

This section calculates GDP per capita from the two existing columns, and derives a `Continent` column from each country's `Country Code` (ISO3). Before calculating GDP per capita, both columns need to be checked and converted to proper numeric types, since they were read as strings from the raw World Bank export.

In [28]:
cols_to_convert = ["GDP (current US$)", "Population, total"]

merged_df = cleaning.convert_columns_to_numeric(merged_df, cols_to_convert)

In [29]:
merged_df["GDP per capita (current US$)"] = (
    merged_df["GDP (current US$)"] / merged_df["Population, total"]
)

merged_df.head()

,Country Name,Country Code,Year,Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),"Foreign direct investment, net inflows (% of GDP)",GDP (current US$),GDP growth (annual %),"Inflation, consumer prices (annual %)","Life expectancy at birth, total (years)","Population, total",...,Tax Burden,Government Spending,Fiscal Health,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom,GDP per capita (current US$)
0,Afghanistan,AFG,2001,0.9402,0.0241685679088889,2.813572e+09,-9.4319740700862,..,55.511,20284307,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,138.706822
1,Afghanistan,AFG,2002,0.9388,1.30694987042884,3.825701e+09,28.6000011706788,..,56.225,21378117,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,178.954088
2,Afghanistan,AFG,2003,1.0172,1.27849325196424,4.520947e+09,8.83227780288267,..,57.171,22733049,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,198.871116
3,Afghanistan,AFG,2004,0.9024,3.577104200584,5.224897e+09,1.41411799339429,..,57.81,23560654,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,221.763654
4,Afghanistan,AFG,2005,1.2763,4.36867310434287,6.203257e+09,11.2297148272859,12.6862687216715,58.247,24404567,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,254.184249


In [30]:
merged_df = cleaning.add_continent_column(merged_df)

merged_df["Continent"].value_counts(dropna=False)

Continent
Africa           1315
Asia             1202
Europe           1014
North America     500
South America     300
Oceania           250
Name: count, dtype: int64

## Part VI | Handling Missing Values

Even after merging two internally-consistent datasets, individual indicators may still contain missing values (not every country reports every indicator every year), and some columns may contain implausible outliers resulting from data entry or reporting inconsistencies.

This section:
1. Quantifies missing values per column and identifies patterns (e.g. concentrated in specific years, countries, or indicators)
2. Decides, indicator by indicator, whether to drop the column, keep it as-is, or correct clearly erroneous values
3. Records the reasoning behind each decision, so that a column's absence is traceable rather than assumed

In [31]:
merged_df.isna().sum()

Country Name                                                              0
Country Code                                                              0
Year                                                                      0
Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e)         0
Foreign direct investment, net inflows (% of GDP)                         0
GDP (current US$)                                                        59
GDP growth (annual %)                                                     0
Inflation, consumer prices (annual %)                                     0
Life expectancy at birth, total (years)                                   0
Population, total                                                         0
Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population)       0
School enrollment, secondary (% gross)                                    0
Unemployment, total (% of total labor force) (modeled ILO estimate)       0
Country     

In [32]:
missing_pct = (merged_df.isna().sum() / len(merged_df) * 100).sort_values(
    ascending=False
)
missing_pct

Fiscal Health                                                          64.833006
Judicial Effectiveness                                                 64.221786
Labor Freedom                                                          19.537219
Overall Score                                                           7.116350
Tax Burden                                                              6.745252
Financial Freedom                                                       6.701594
Trade Freedom                                                           6.505130
Government Spending                                                     6.417813
Monetary Freedom                                                        6.243178
Investment Freedom                                                      6.221349
Business Freedom                                                        5.893910
Property Rights                                                         5.850251
Government Integrity        

In [33]:
merged_df = merged_df.drop(
    columns=[
        "Index Year",
        "Fiscal Health",
        "Judicial Effectiveness",
        "Country",
    ]
)

merged_df.head()

,Country Name,Country Code,Year,Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),"Foreign direct investment, net inflows (% of GDP)",GDP (current US$),GDP growth (annual %),"Inflation, consumer prices (annual %)","Life expectancy at birth, total (years)","Population, total",...,Tax Burden,Government Spending,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom,GDP per capita (current US$),Continent
0,Afghanistan,AFG,2001,0.9402,0.0241685679088889,2.813572e+09,-9.4319740700862,..,55.511,20284307,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,138.706822,Asia
1,Afghanistan,AFG,2002,0.9388,1.30694987042884,3.825701e+09,28.6000011706788,..,56.225,21378117,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,178.954088,Asia
2,Afghanistan,AFG,2003,1.0172,1.27849325196424,4.520947e+09,8.83227780288267,..,57.171,22733049,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,198.871116,Asia
3,Afghanistan,AFG,2004,0.9024,3.577104200584,5.224897e+09,1.41411799339429,..,57.81,23560654,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,221.763654,Asia
4,Afghanistan,AFG,2005,1.2763,4.36867310434287,6.203257e+09,11.2297148272859,12.6862687216715,58.247,24404567,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,254.184249,Asia


In [34]:
merged_df.shape


(4581, 26)

In [35]:
merged_df.columns.tolist()

['Country Name',
 'Country Code',
 'Year',
 'Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e)',
 'Foreign direct investment, net inflows (% of GDP)',
 'GDP (current US$)',
 'GDP growth (annual %)',
 'Inflation, consumer prices (annual %)',
 'Life expectancy at birth, total (years)',
 'Population, total',
 'Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population)',
 'School enrollment, secondary (% gross)',
 'Unemployment, total (% of total labor force) (modeled ILO estimate)',
 'Overall Score',
 'Property Rights',
 'Government Integrity',
 'Tax Burden',
 'Government Spending',
 'Business Freedom',
 'Labor Freedom',
 'Monetary Freedom',
 'Trade Freedom',
 'Investment Freedom',
 'Financial Freedom',
 'GDP per capita (current US$)',
 'Continent']

In [36]:
missing_pct = (merged_df.isna().sum() / len(merged_df) * 100).sort_values(
    ascending=False
)
missing_pct

Labor Freedom                                                          19.537219
Overall Score                                                           7.116350
Tax Burden                                                              6.745252
Financial Freedom                                                       6.701594
Trade Freedom                                                           6.505130
Government Spending                                                     6.417813
Monetary Freedom                                                        6.243178
Investment Freedom                                                      6.221349
Business Freedom                                                        5.893910
Property Rights                                                         5.850251
Government Integrity                                                    5.500982
GDP (current US$)                                                       1.287928
GDP per capita (current US$)

### Investigating the Labor Freedom Gap

In [37]:
merged_df[merged_df["Labor Freedom"].isna()][
    "Year"
].value_counts().sort_index()


Year
2001    182
2002    182
2003    182
2004    182
2005     28
2006     26
2007     26
2008     26
2009      4
2010      4
2011      5
2012      6
2013      3
2014      2
2015      2
2016      2
2018      2
2019      2
2020      2
2021      2
2022      7
2023      2
2024      8
2025      8
Name: count, dtype: int64

In [38]:
merged_df[merged_df["Labor Freedom"].isna()][
    "Country Name"
].value_counts().head(15)

Country Name
Liechtenstein                   24
Afghanistan                     15
Iraq                            15
Somalia                         13
Sudan                           12
Bhutan                           8
Comoros                          8
Democratic Republic of Congo     8
Dominica                         8
Eritrea                          8
Kiribati                         8
Liberia                          8
Libya                            8
Macau                            8
Maldives                         8
Name: count, dtype: int64

**Conclusion:** Labor Freedom has 19.5% missing values, heavily concentrated in 2001–2008 (up to 182 out of ~185 countries missing per year), dropping to near-zero from 2009 onwards. This reflects the component's later introduction into the Index methodology, rather than a data collection issue. The column is retained, but any time-series analysis involving Labor Freedom should account for its limited historical coverage before 2009.

The remaining missing values across other columns are concentrated in a small number of countries with historically limited data availability (e.g. Liechtenstein, Afghanistan, Iraq, Somalia, Bhutan), consistent with instability, small population, or limited institutional reporting capacity.

---

### Casting Columns to Their Proper Numeric Type

`.describe()` above covers only a handful of columns, because the World Bank export delivers its indicators as strings: thousands separators make them non-numeric, and `..` is used as the null marker. Both are converted below.

In [39]:
merged_df.describe()

,Year,GDP (current US$),"Population, total",Overall Score,Property Rights,Government Integrity,Tax Burden,Government Spending,Business Freedom,Labor Freedom,Monetary Freedom,Trade Freedom,Investment Freedom,Financial Freedom,GDP per capita (current US$)
count,4581.000000,4.522000e+03,4.581000e+03,4255.000000,4313.000000,4329.000000,4272.000000,4287.000000,4311.000000,3686.000000,4295.000000,4283.000000,4296.000000,4274.000000,4522.000000
mean,2013.028378,4.037714e+11,3.930915e+07,59.931469,48.227359,41.721113,75.715824,64.621833,63.274832,59.680846,74.079348,71.475835,53.634777,49.493449,13786.065434
std,7.201594,1.743695e+12,1.424973e+08,11.369350,23.834680,21.899628,14.085511,23.882483,16.488645,15.395355,12.405376,13.841758,21.884010,19.719709,22128.741705
min,2001.000000,6.493585e+07,3.319400e+04,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,109.593814
25%,2007.000000,8.034019e+09,2.376722e+06,53.200000,30.000000,26.000000,68.700000,51.300000,54.800000,50.000000,70.600000,64.000000,40.000000,30.000000,1425.883236
50%,2013.000000,3.074509e+10,8.720283e+06,59.700000,46.000000,35.600000,77.500000,70.400000,63.900000,59.600000,76.100000,73.600000,55.000000,50.000000,4556.230443
75%,2019.000000,1.926441e+11,2.764936e+07,67.400000,66.700000,53.000000,84.725000,83.400000,73.900000,69.700000,81.000000,81.400000,70.000000,60.000000,15659.163297
max,2025.000000,3.076970e+13,1.463866e+09,90.200000,100.000000,100.000000,100.000000,99.300000,100.000000,100.000000,95.400000,95.000000,95.000000,90.000000,220167.225474


In [40]:
merged_df = cleaning.convert_columns_to_numeric(
    merged_df, cleaning.WB_NUMERIC_COLS
)

In [41]:
merged_df.describe().T

,count,mean,std,min,25%,50%,75%,max
Year,4581.0,2.013028e+03,7.201594e+00,2.001000e+03,2.007000e+03,2.013000e+03,2.019000e+03,2.025000e+03
Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),4313.0,1.820587e+02,8.556816e+02,2.000000e-04,2.984300e+00,1.409680e+01,7.294110e+01,1.312473e+04
"Foreign direct investment, net inflows (% of GDP)",4256.0,7.311693e+00,4.974529e+01,-1.303108e+03,1.190340e+00,2.809549e+00,5.594378e+00,1.282607e+03
GDP (current US$),4522.0,4.037714e+11,1.743695e+12,6.493585e+07,8.034019e+09,3.074509e+10,1.926441e+11,3.076970e+13
GDP growth (annual %),4498.0,3.530866e+00,5.533162e+00,-5.440209e+01,1.501833e+00,3.664719e+00,5.879458e+00,8.682675e+01
"Inflation, consumer prices (annual %)",4263.0,6.459680e+00,1.667305e+01,-1.685969e+01,1.731484e+00,3.604522e+00,7.189209e+00,5.572018e+02
"Life expectancy at birth, total (years)",4398.0,7.045339e+01,8.702854e+00,1.466500e+01,6.493150e+01,7.190037e+01,7.700300e+01,8.549634e+01
"Population, total",4581.0,3.930915e+07,1.424973e+08,3.319400e+04,2.376722e+06,8.720283e+06,2.764936e+07,1.463866e+09
Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population),1784.0,9.035258e+00,1.666562e+01,0.000000e+00,3.000000e-01,1.300000e+00,8.925000e+00,9.490000e+01
"School enrollment, secondary (% gross)",3177.0,8.274280e+01,2.850791e+01,3.347850e+00,6.709386e+01,8.975031e+01,1.014880e+02,1.640798e+02


### Re-checking Missing Values After Numeric Conversion

The `missing_pct` check above ran before the `".."` → `NaN` conversion in `convert_columns_to_numeric()`, so every raw World Bank column that used that null marker showed up as 0% missing. Re-running it now, after conversion, gives the real picture.

In [42]:
missing_pct = (merged_df.isna().sum() / len(merged_df) * 100).sort_values(
    ascending=False
)
missing_pct

Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population)    61.056538
School enrollment, secondary (% gross)                                 30.648330
Labor Freedom                                                          19.537219
Overall Score                                                           7.116350
Foreign direct investment, net inflows (% of GDP)                       7.094521
Inflation, consumer prices (annual %)                                   6.941716
Tax Burden                                                              6.745252
Financial Freedom                                                       6.701594
Trade Freedom                                                           6.505130
Government Spending                                                     6.417813
Monetary Freedom                                                        6.243178
Investment Freedom                                                      6.221349
Business Freedom            

### Investigating the Poverty Headcount Ratio Column

Poverty headcount ratio at $3.00 a day now shows 61.1% missing, by far the sparsest column in the dataset, well beyond Labor Freedom's 19.5%.

In [43]:
pov_col = "Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population)"

merged_df[merged_df[pov_col].isna()]["Year"].value_counts().sort_index()

Year
2001    134
2002    113
2003    113
2004    107
2005    103
2006    102
2007    111
2008    103
2009    102
2010     99
2011    103
2012     98
2013    107
2014    101
2015     98
2016    104
2017    107
2018     93
2019    108
2020    115
2021    102
2022    110
2023    127
2024    158
2025    179
Name: count, dtype: int64

In [44]:
(
    merged_df.groupby("Country Name")[pov_col].apply(
        lambda s: s.notna().sum()
    )
    == 0
).sum()

np.int64(23)

In [45]:
merged_df[merged_df[pov_col].isna()]["Country Name"].value_counts().head(15)

Country Name
Afghanistan                         25
Bahrain                             25
Cambodia                            25
Cuba                                25
Dominica                            25
Eritrea                             25
Guyana                              25
Kuwait                              25
Libya                               25
Liechtenstein                       25
New Zealand                         25
North Korea                         25
Oman                                25
Saint Vincent and the Grenadines    25
Saudi Arabia                        25
Name: count, dtype: int64

**Conclusion:** Missingness is spread fairly evenly across all 25 years (roughly half of country-years missing each year, worse in 2023–2025 due to reporting lag), and 23 of the 185 countries never report this indicator at all: a mix of conflict-affected or low-capacity states without regular survey infrastructure (Afghanistan, Libya, Eritrea, North Korea) and wealthy states where $3/day poverty is effectively zero and is not tracked (Kuwait, New Zealand, Liechtenstein, Saudi Arabia). This reflects how infrequently household consumption surveys are run, not a data error.

Given over 60% missing and zero coverage for more than an eighth of the countries in the dataset, this column is not reliable enough to keep for correlation work, so it was dropped.

---

In [46]:
merged_df = merged_df.drop(columns=[pov_col])

merged_df.shape

(4581, 25)

### Investigating the School Enrollment Column

School enrollment, secondary now shows 30.6% missing.

In [47]:
sch_col = "School enrollment, secondary (% gross)"

(
    merged_df[merged_df[sch_col].isna()]["Year"].value_counts().sort_index()
    / merged_df["Year"].value_counts().sort_index()
    * 100
).round(1)

Year
2001    26.9
2002    26.9
2003    29.1
2004    24.2
2005    23.1
2006    26.4
2007    24.7
2008    26.4
2009    27.5
2010    24.2
2011    21.9
2012    25.5
2013    29.3
2014    26.5
2015    22.7
2016    27.0
2017    28.6
2018    30.8
2019    30.3
2020    29.7
2021    27.3
2022    32.2
2023    29.0
2024    50.3
2025    95.6
Name: count, dtype: float64

In [48]:
merged_df[merged_df[sch_col].isna()]["Country Name"].value_counts().head(15)

Country Name
Djibouti               25
Haiti                  25
North Korea            24
Zambia                 24
Guinea-Bissau          23
Micronesia             23
Uganda                 23
Equatorial Guinea      22
Gabon                  22
Trinidad and Tobago    22
Iraq                   21
Liberia                21
Libya                  21
Republic of Congo      21
The Gambia             21
Name: count, dtype: int64

**Conclusion:** Missingness sits around 22–30% for most years but spikes sharply in the most recent two (50.3% in 2024 and 95.6% in 2025), consistent with reporting lag rather than a structural gap, since secondary enrollment figures for the current year are rarely available yet. The remaining missingness is concentrated in a handful of countries with historically incomplete education reporting (Djibouti, Haiti, North Korea, Zambia, Guinea-Bissau).

This is far less severe than the poverty ratio and largely explainable, so the column was kept. Limitation for future work: analyses using recent years, especially 2024–2025, should account for near-total missingness in this column in those years.

---

## Part VII | Investigating Outliers

Extreme values are not errors by default. This section takes each indicator with an implausible-looking tail and asks the same question: does the value correspond to a documented real-world event, or does it have no possible real-world reading?

Only the second kind is corrected. Everything else is kept, because smoothing a genuine crisis out of the record would misrepresent exactly the periods this dataset is most useful for.

### Life Expectancy

In [49]:
merged_df.nsmallest(5, "Life expectancy at birth, total (years)")[
    ["Country Name", "Year", "Life expectancy at birth, total (years)"]
]

,Country Name,Year,"Life expectancy at birth, total (years)"
745,Central African Republic,2009,14.665
758,Central African Republic,2022,18.818
755,Central African Republic,2019,31.530
3666,Somalia,2011,32.453
750,Central African Republic,2014,40.265


In [50]:
merged_df[(merged_df["Country Name"] == "Central African Republic")][
    "Life expectancy at birth, total (years)"
].sort_index()

737    46.734
738    46.840
739    47.306
740    47.696
741    48.207
742    48.687
743    49.327
744    49.873
745    14.665
746    46.871
747    50.107
748    51.336
749    49.487
750    40.265
751    51.934
752    50.999
753    45.231
754    52.297
755    31.530
756    50.596
757    40.279
758    18.818
759    57.408
760    57.671
761       NaN
Name: Life expectancy at birth, total (years), dtype: float64

In [51]:
merged_df.loc[
    merged_df["Life expectancy at birth, total (years)"] < 20,
    "Life expectancy at birth, total (years)",
] = pd.NA

**Conclusion:** The five lowest life expectancy values in the dataset are dominated by the Central African Republic, including two implausible entries: 14.665 years (2009) and 18.818 years (2022). No country has ever recorded a national life expectancy this low, even during severe humanitarian crises; the rest of the CAR series sits in the 30–55 year range, confirming these two values are isolated data errors rather than a genuine trend.

These values were replaced with `NaN`, treating them as missing rather than allowing clearly invalid data to distort the downstream analysis.

---

### Foreign Direct Investment

In [52]:
merged_df.nsmallest(10, "Foreign direct investment, net inflows (% of GDP)")[
    [
        "Country Name",
        "Year",
        "Foreign direct investment, net inflows (% of GDP)",
    ]
]

,Country Name,Year,"Foreign direct investment, net inflows (% of GDP)"
2338,Liechtenstein,2018,-1303.108267
2392,Luxembourg,2022,-391.555119
981,Cyprus,2020,-296.013333
985,Cyprus,2024,-133.824213
2388,Luxembourg,2018,-117.233948
984,Cyprus,2023,-63.154639
2377,Luxembourg,2007,-57.532314
2387,Luxembourg,2017,-41.650995
1749,Hungary,2018,-40.108246
2706,Mongolia,2016,-37.172653


In [53]:
merged_df.nlargest(10, "Foreign direct investment, net inflows (% of GDP)")[
    [
        "Country Name",
        "Year",
        "Foreign direct investment, net inflows (% of GDP)",
    ]
]

,Country Name,Year,"Foreign direct investment, net inflows (% of GDP)"
2335,Liechtenstein,2015,1282.607229
2326,Liechtenstein,2006,1114.932386
2336,Liechtenstein,2016,972.693539
2327,Liechtenstein,2007,728.865660
2329,Liechtenstein,2009,664.573767
2337,Liechtenstein,2017,602.374526
2330,Liechtenstein,2010,587.838616
2323,Liechtenstein,2003,563.548408
2332,Liechtenstein,2012,540.366946
2331,Liechtenstein,2011,460.213518


In [54]:
extreme_fdi = merged_df[
    merged_df["Foreign direct investment, net inflows (% of GDP)"].abs() > 100
]
extreme_fdi["Country Name"].value_counts()

Country Name
Liechtenstein    15
Malta            14
Cyprus           12
Luxembourg        4
Hungary           1
Liberia           1
Name: count, dtype: int64

**Conclusion:** The extreme FDI values are not isolated to Liechtenstein. Restricting to rows where FDI inflows exceed ±100% of GDP, the vast majority cluster in four countries: Liechtenstein (15), Malta (14), Cyprus (12), and Luxembourg (4), all small, financially open economies with international banking or holding-company sectors that are large relative to their domestic GDP. Hungary and Liberia each cross the threshold only once, consistent with one-off events rather than a structural pattern. In the four financial-centre economies, flows tied to special purpose entities and pass-through capital can dwarf the real economy, producing FDI-to-GDP swings of several hundred percent in a single year; this is a documented feature of how FDI is recorded for these economies, not a data quality error.

These values were kept as-is, consistent with the treatment of the GDP growth and inflation outliers above. However, this concentration is worth flagging as a limitation: a handful of countries dominate any FDI aggregate, so a correlation involving this column is largely a statement about four financial centres unless they are winsorised, log-scaled, or flagged separately.

---

### GDP Growth

In [55]:
merged_df.nsmallest(5, "GDP growth (annual %)")[
    ["Country Name", "Year", "GDP growth (annual %)"]
]

,Country Name,Year,GDP growth (annual %)
2415,Macau,2020,-54.402093
2306,Libya,2011,-50.338515
1859,Iraq,2003,-36.656780
749,Central African Republic,2013,-36.391977
2510,Maldives,2020,-32.908829


In [56]:
merged_df.nlargest(5, "GDP growth (annual %)")[
    ["Country Name", "Year", "GDP growth (annual %)"]
]

,Country Name,Year,GDP growth (annual %)
2307,Libya,2012,86.826748
1237,Equatorial Guinea,2001,63.379875
1658,Guyana,2022,63.334634
1860,Iraq,2004,53.385547
1660,Guyana,2024,43.819273


**Conclusion:** Both ends of the GDP growth distribution are dominated by real macroeconomic shocks rather than data errors. The sharpest contractions, Macau (-54.4%, 2020) and the Maldives (-32.9%, 2020), are COVID-19 tourism collapses in economies that depend heavily on international visitors; Libya (-50.3%, 2011), Iraq (-36.7%, 2003), and the Central African Republic (-36.4%, 2013) coincide exactly with civil war, invasion, or coup years in each country. The sharpest expansions mirror the same events from the other side: Libya (+86.8%, 2012) and Iraq (+53.4%, 2004) are post-conflict rebounds off a collapsed prior-year base, while Equatorial Guinea (+63.4%, 2001) and Guyana (+63.3% in 2022, +43.8% in 2024) reflect the start of major oil production.

These values were kept as-is: each one is tied to a specific, verifiable historical event rather than a measurement artefact, so removing or capping them would discard genuine economic signal.

---

### Inflation

In [57]:
merged_df.nlargest(5, "Inflation, consumer prices (annual %)")[
    ["Country Name", "Year", "Inflation, consumer prices (annual %)"]
]

,Country Name,Year,"Inflation, consumer prices (annual %)"
4575,Zimbabwe,2020,557.201817
1037,Democratic Republic of Congo,2001,359.936614
3801,Sudan,2021,359.093041
4574,Zimbabwe,2019,255.304991
4471,Venezuela,2016,254.948535


**Conclusion:** The highest inflation readings in the dataset are Zimbabwe (557%, 2020), the Democratic Republic of Congo (360%, 2001), Sudan (359%, 2021), Zimbabwe again (255%, 2019), and Venezuela (255%, 2016), all corresponding to well-documented hyperinflation episodes driven by currency collapse, civil conflict, or economic mismanagement in those specific years. This matches the broader pattern already seen in the GDP growth outliers: extreme values cluster around known crises rather than appearing at random.

These values were kept as-is, consistent with the treatment of the other outliers investigated in this section: they are real economic events, and smoothing or removing them would misrepresent the historical record.

---

## Part VIII | Normalising Numeric Precision

Every value in `merged_df` is correct by now, but not every value is *readable*. The World Bank export and the derived columns carry up to eighteen decimal places — an inflation rate quoted to 1e-16 of a percentage point, a GDP figure quoted well past the cent — which is precision neither source possesses, and which makes the exported CSV impossible to scan by eye.

This step pins each numeric column to the precision its source actually supports, through the `cleaning.NUMERIC_PRECISION` mapping: rates, percentages and per-capita figures to two decimals, absolute magnitudes to whole units, and the Heritage sub-scores to the single decimal Heritage itself publishes. CO2 emissions are the one exception at four decimals: 24 country-years report less than 0.005 Mt (the smallest, 0.0002 Mt), and two decimals would flatten every one of them to zero.

Rounding is deliberately the *only* rescaling applied here. Standardising or min-max scaling the columns would destroy exactly the readability the step exists for, and would leave the published table in units no reader can interpret — a GDP per capita of `-0.43` means nothing on its own. The export keeps its natural units.

In [58]:
cleaning.decimal_places(merged_df)

Inflation, consumer prices (annual %)                                  18
Foreign direct investment, net inflows (% of GDP)                      16
GDP growth (annual %)                                                  16
School enrollment, secondary (% gross)                                 14
GDP per capita (current US$)                                           14
Life expectancy at birth, total (years)                                13
GDP (current US$)                                                       7
Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e)       4
Unemployment, total (% of total labor force) (modeled ILO estimate)     3
Monetary Freedom                                                        1
Government Integrity                                                    1
Property Rights                                                         1
Tax Burden                                                              1
Overall Score                         

In [59]:
# Rounding is idempotent, so holding on to the pre-rounding frame for the
# verification below survives a re-run of this cell unchanged.
unrounded_df = merged_df

merged_df = cleaning.round_numeric_columns(merged_df)

In [60]:
cleaning.decimal_places(merged_df)

Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e)      4
Foreign direct investment, net inflows (% of GDP)                      2
Inflation, consumer prices (annual %)                                  2
GDP growth (annual %)                                                  2
School enrollment, secondary (% gross)                                 2
Life expectancy at birth, total (years)                                2
Unemployment, total (% of total labor force) (modeled ILO estimate)    2
GDP per capita (current US$)                                           2
Monetary Freedom                                                       1
Government Spending                                                    1
Government Integrity                                                   1
Property Rights                                                        1
Tax Burden                                                             1
Overall Score                                      

In [61]:
# No value should move by more than half a unit in the last kept decimal
# place. Worth confirming rather than assuming, since this is the one step
# that alters the numbers themselves.
numeric_cols = merged_df.select_dtypes("number").columns

(
    (merged_df[numeric_cols] - unrounded_df[numeric_cols])
    .abs()
    .max()
    .sort_values(ascending=False)
)

GDP (current US$)                                                      0.500000
Life expectancy at birth, total (years)                                0.005000
Unemployment, total (% of total labor force) (modeled ILO estimate)    0.005000
GDP growth (annual %)                                                  0.005000
Inflation, consumer prices (annual %)                                  0.005000
School enrollment, secondary (% gross)                                 0.005000
Foreign direct investment, net inflows (% of GDP)                      0.005000
GDP per capita (current US$)                                           0.004999
Year                                                                   0.000000
Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e)      0.000000
Overall Score                                                          0.000000
Property Rights                                                        0.000000
Government Integrity                    

In [62]:
merged_df.head(3).T

,0,1,2
Country Name,Afghanistan,Afghanistan,Afghanistan
Country Code,AFG,AFG,AFG
Year,2001,2002,2003
Carbon dioxide (CO2) emissions (total) excluding LULUCF (Mt CO2e),0.9402,0.9388,1.0172
"Foreign direct investment, net inflows (% of GDP)",0.02,1.31,1.28
GDP (current US$),2813571754.0,3825701439.0,4520946819.0
GDP growth (annual %),-9.43,28.6,8.83
"Inflation, consumer prices (annual %)",NaN,NaN,NaN
"Life expectancy at birth, total (years)",55.51,56.22,57.17
"Population, total",20284307,21378117,22733049


**Conclusion:** The table now carries at most four decimal places instead of eighteen, and the exported file shrinks by 29%. The largest change absorbed by any single value is half a unit in the last kept place — half a dollar on GDP, half a cent on GDP per capita, 0.005 of a percentage point on every rate — so the rounding is cosmetic against every real signal in the data: the rank ordering within each column is preserved (Spearman ≥ 0.999999), the missingness pattern is untouched, and the crisis outliers kept in Part VII survive intact.

Two consequences are worth recording. First, thirteen near-zero rate readings (nine FDI, two GDP growth, two inflation) now sit at exactly `0.00`; they were never distinguishable from zero at the sources' true accuracy, but a log transform would still need them handled. Second, the `GDP per capita = GDP / Population` identity now holds to a relative error of 4.0e-5 rather than 2.5e-16 — bounded by this rounding rather than by float arithmetic, and still exact enough that the three columns carry two independent quantities, not three.

## Part IX | Exporting the Cleaned Dataset

With country standardisation, reshaping, merging, missing-value handling, outlier review, and precision normalisation all complete, `merged_df` is now the final cleaned, analysis-ready dataset. It is exported to `data/processed/` so that `II-analysis.ipynb` can load it directly, without re-running the full cleaning pipeline.

In [63]:
print(f"Final shape: {merged_df.shape[0]} rows x {merged_df.shape[1]} columns")
print(f"Countries: {merged_df['Country Name'].nunique()}")
print(f"Years: {merged_df['Year'].min()}-{merged_df['Year'].max()}")

Final shape: 4581 rows x 25 columns
Countries: 185
Years: 2001-2025


In [64]:
merged_df.to_csv(
    "data/processed/wb_economic_freedom_merged.csv", index=False
)